### PR TITLE
Add embeddings for LocalAI

### PR DESCRIPTION
**Pull request by mudler**
Description:

This PR adds embeddings for LocalAI ( https://github.com/go-skynet/LocalAI ), a self-hosted OpenAI drop-in replacement. As LocalAI can re-use OpenAI clients it is mostly following the lines of the OpenAI embeddings, however when embedding documents, it just uses string instead of sending tokens as sending tokens is best-effort depending on the model being used in LocalAI. Sending tokens is also tricky as token id's can mismatch with the model - so it's safer to just send strings in this case.

Partly related to: https://github.com/hwchase17/langchain/issues/5256

Dependencies: No new dependencies

Twitter: @mudler_it

Maintainers: @rlancemartin, @eyurtsev, @hwchase17

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
1. lets add an example notebook for this
2. lets add this to `langchain/embeddings/__init__.py`

**Comment by mudler** [comment]
@hwchase17 done! will follow up along with #5256 and once we have complete integration with LocalAI I'll update also the documentation page accordingly. 

Taking this opportunity to ask - is there any interest into adding e.g. voice capabilities? LocalAI supports tts and audio-to-text as well

**Comment by baskaryan** [comment]
looks awesome, thanks @mudler!

there definitely is interest in voice but adding other modalities is a big change that we want to be super thoughtful about, and we haven't had the time to think it through just yet. very open to suggestions on the interface if you're eager to see it in langchain

File: libs/langchain/langchain/embeddings/localai.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+            raise AssertionError("this is unreachable")
+
+        return wrapped_f
+
+    return wrap
+
+
+# https://stackoverflow.com/questions/76469415/getting-embeddings-of-length-1-from-langchain-openaiembeddings
+def _check_response(response: dict) -> dict:
+    if any(len(d["embedding"]) == 1 for d in response["data"]):
```
Comment (mkhludnev) [inline libs/langchain/langchain/embeddings/localai.py:90]: Pardon for commenting old code. But the  SO thread discusses OpenAI service, so it's hardly the case in LocalAI. Isn't it? 
This is why I want to remove this retry condition in new integration package https://github.com/mkhludnev/langchain-localai WDYT?
This package spin off from the discussion https://github.com/langchain-ai/langchain/pull/22399#issuecomment-2537387476

File: libs/langchain/langchain/embeddings/localai.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+                "http": self.openai_proxy,
+                "https": self.openai_proxy,
+            }  # type: ignore[assignment]  # noqa: E501
+        return openai_args
+
+    def _embedding_func(self, text: str, *, engine: str) -> List[float]:
+        """Call out to LocalAI's embedding endpoint."""
+        # handle large input text
+        if self.model.endswith("001"):
+            # See: https://github.com/openai/openai-python/issues/418#issuecomment-1525939500
```
Comment (mkhludnev) [inline libs/langchain/langchain/embeddings/localai.py:262]: Should we bother about OpenAI specifics when integrate with LocalAI? I don't think so. I'm going to wipe it there https://github.com/mkhludnev/langchain-localai

**Pull request merged**